### ISSUE TITLE
documenting range properties of the views (input and output)

### ISSUE THREAD

**Issue by h-2** [ericniebler/range-v3#701]
I would like to add documentation that covers which range properties are preserved by views and which are lost (and possibly which are added).

If you think this is a good idea, I would first create a table in this issue, double-check the content with you, and then create PRs for the documentation. That is, if you agree this would be nice to have.

I though about something like this:

|                    |  `input_range`  | `forward_range` | `bidirectional_range` | `random_access_range` | `sized_range` |
|--------------------|:---------------:|:---------------:|:---------------------:|:---------------------:|:-------------:|
|`view::join`        |      ☑          | `input_range`   |  `input_range`        |   `input_range`       |      ☐        |
|`view::transform`   |      ☑          |    ☑            |      ☑                |      ☑                |      ☑        |

**Comment by h-2** [comment]
I went ahead and created a table in the wiki, I hope that was ok:
https://github.com/ericniebler/range-v3/wiki/Range-property-requirements-and-conservation-by-views

I will try to take some time off and fill up some values, but everyone else is invited to contribute, as well 😃

**Comment by ericniebler** [comment]
Good idea.

**Comment by h-2** [comment]
@ericniebler I was just adding the information for `view::take_while` and was surprised that it only loses size, but preserves random_access. [] past the predicate-induced end actually "works" and resolves to positions behind the end on the underlying range. I assume this is a bug?

**Comment by ericniebler** [comment]
On one hand, sure, it is sensible that `take_while` should be Forward or Bidirectional at best. OTOH, I'm of the opinion that a null-terminated string (say) should be random-access. If you use the random-access-ness to jump past the null terminator, the fault is yours. There are examples of algorithms that consume a range safely one element at a time, and then want random access into the part that has already been traversed. (Backtracking string searching algorithms come to mind.) The `take_while` view is another example of the same principle.

TL;DR, I think this is intentional.

**Comment by h-2** [comment]
> OTOH, I'm of the opinion that a null-terminated string (say) should be random-access. If you use the random-access-ness to jump past the null terminator, the fault is yours.

Hm, but this is ugly c-legacy, no? 😱 I mean, I agree that the [] operator shouldn't do bounds checking, but should we promote it's use if we know there *cannot* be bounds-checking? c-string interfaces explicitly carry around a size "partner" specifically for this reason, but for `view::take_while` this doesn't hold.

> There are examples of algorithms that consume a range safely one element at a time, and then want random access into the part that has already been traversed. (Backtracking string searching algorithms come to mind.)

This OTOH is something I didn't think about and I can see it's usefulness💡 However you could achieve this flexibility if you call `view::take_exactly(current_position)` on your `view::take_while`, because that will re-introduce size. Then again, re-introducing random-access will be more difficult, right?
But, I do think, it would be cleaner/safer.

**Comment by ericniebler** [comment]
> Hm, but this is ugly c-legacy, no? 😱 I mean, I agree that the [] operator shouldn't do bounds checking, but should we promote it's use if we know there _cannot_ be bounds-checking? 

Random access iterators provide operator [] and operator+ despite the fact that in general they cannot be bounds-checked. The STL, and C++ in general, doesn't stop you from walking off the ends of ranges, whether they're arrays, vectors, strings, lists, or whatever.

You might draw the line with ranges and say that their interfaces should be safer than the necessarily more low-level iterators. That is, `rng[n]` might fail to compile for some ranges while `rng.begin()[n]` works. This feels like too much nannying to me, but regardless it's a decision that someone might make for their range. (The `Range` concept has nothing to say about `operator[]`.) 

I like op[] for range types with random-access iterators, and I prefer the consistency of having `rng[n]` and `begin(rng)[n]` being synonymous.

> Then again, re-introducing random-access will be more difficult, right?

Impossible in principle, I'd say. There's no mechanism to ask an iterator if it's really a random-access iterator hidden inside a forward iterator.

**Comment by h-2** [comment]
> Impossible in principle, I'd say. There's no mechanism to ask an iterator if it's really a random-access iterator hidden inside a forward iterator.

Hm, I agree, it could just be done with specific hacks / or would drastically complicate interfaces. I still don't like it, but I don't see a better solution right now. Thanks for taking the time to explain your point of view!
